pkg/operator/controller/ingress_status.go

```diff
@@ -0,0 +1,102 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// syncIngressControllerStatus computes the current status of ic and therefrom
+// updates the syncIngressControllerStatus resource.
+func (r *reconciler) syncIngressControllerStatus(deployment *appsv1.Deployment, ic *operatorv1.IngressController) error {
+	selector, err := metav1.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return fmt.Errorf("ingress controller deployment %s/%s has invalid spec.selector: %v", deployment.Namespace, deployment.Name, err)
+	}
+
+	updated := ic.DeepCopy()
+	updated.Status.AvailableReplicas = deployment.Status.AvailableReplicas
+	updated.Status.Selector = selector.String()
+	updated.Status.Conditions = computeIngressStatusConditions(updated.Status.Conditions, deployment)
+	if !ingressStatusesEqual(updated.Status, ic.Status) {
+		if err := r.client.Status().Update(context.TODO(), updated); err != nil {
+			return fmt.Errorf("failed to update the status of IngressController %s/%s", ic.Namespace, ic.Name)
+		}
+	}
+
+	return nil
+}
+
+// computeIngressStatusConditions computes the ingress controller's current state.
+func computeIngressStatusConditions(conditions []operatorv1.OperatorCondition, deployment *appsv1.Deployment) []operatorv1.OperatorCondition {
+	availableCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.IngressControllerAvailableConditionType,
+		Status: operatorv1.ConditionUnknown,
+	}
+	if deployment.Status.AvailableReplicas > 0 {
+		availableCondition.Status = operatorv1.ConditionTrue
+	} else {
+		availableCondition.Status = operatorv1.ConditionFalse
+		availableCondition.Reason = "IngressControllerUnavailable"
```

Comment:
Should this be in terms of the deployment? e.g. `DeploymentUnavailable` and `%d deployment replicas available, want %d`?